File: apps/mobile/android/app/src/main/java/com/beacon/bluetooth/mapping/ConnectionStateMapper.kt
package com.beacon.bluetooth.mapping

import android.bluetooth.BluetoothProfile

/**
 * Connection vocabulary shared with TypeScript (`ConnectionState` in `@beacon/ble-contracts`).
 * Wire values must match that union exactly.
 */
enum class BleConnectionState(val wireValue: String) {
    DISCONNECTED("disconnected"),
    CONNECTING("connecting"),
    CONNECTED("connected"),
    DISCOVERING_SERVICES("discovering_services"),
    READY("ready"),
    DISCONNECTING("disconnecting"),
    FAILED("failed"),
}

/** Pure mapping from `BluetoothProfile` connection states, used for platform callbacks. */
object ConnectionStateMapper {
    fun fromProfileState(state: Int): BleConnectionState =
        when (state) {
            BluetoothProfile.STATE_CONNECTED -> BleConnectionState.CONNECTED
            BluetoothProfile.STATE_CONNECTING -> BleConnectionState.CONNECTING
            BluetoothProfile.STATE_DISCONNECTING -> BleConnectionState.DISCONNECTING
            else -> BleConnectionState.DISCONNECTED
        }
}

File: apps/mobile/ios/BeaconBluetooth/Mapping/ConnectionStateMapper.swift
import CoreBluetooth

/// Connection vocabulary shared with TypeScript (`ConnectionState` in `@beacon/ble-contracts`).
/// Raw values are the wire format. Native emits every transition; JavaScript mirrors them.
public enum BleConnectionState: String, CaseIterable {
  case disconnected
  case connecting
  case connected
  case discoveringServices = "discovering_services"
  case ready
  case disconnecting
  case failed
}

/// Pure mapping from `CBPeripheralState` to the shared vocabulary, used to answer state
/// questions for peripherals this app did not connect itself (e.g. after a relaunch).
public enum ConnectionStateMapper {
  public static func map(_ state: CBPeripheralState) -> BleConnectionState {
    switch state {
    case .disconnected: return .disconnected
    case .connecting: return .connecting
    case .connected: return .connected
    case .disconnecting: return .disconnecting
    @unknown default: return .disconnected
    }
  }
}
